library/core/src/iter/adapters/enumerate.rs

```diff
@@ -15,10 +16,155 @@ use crate::ops::{Add, AddAssign, Try};
+/// Specifically, for slices this allows the optimizer to know that the returned values are never
+/// bigger than the size of the slice.
+///
+/// The only difference between the default and specialized implementation is the use of
+/// `intrinsics::assume()` on the to be returned values, and both implementations must be kept in
+/// sync.
+#[doc(hidden)]
+trait EnumerateImpl<I> {
+    type Item;
+    fn new(iter: I) -> Self;
+    fn next(&mut self) -> Option<Self::Item>;
+    unsafe fn __iterator_get_unchecked(&mut self, idx: usize) -> Self::Item
+    where
+        Self: TrustedRandomAccess;
+    fn next_back(&mut self) -> Option<Self::Item>
+    where
+        I: ExactSizeIterator + DoubleEndedIterator;
+}
+
+impl<I> EnumerateImpl<I> for Enumerate<I>
+where
+    I: Iterator,
+{
+    type Item = (usize, I::Item);
+
+    default fn new(iter: I) -> Self {
+        Enumerate {
+            iter,
+            count: 0,
+            len: 0, // unused
+        }
+    }
+
+    #[inline]
+    default fn next(&mut self) -> Option<Self::Item> {
+        let a = self.iter.next()?;
+        let i = self.count;
+        // Possible undefined overflow. By directly calling the trait method instead of using the
+        // `+=` operator the decision about overflow checking is delayed to the crate that does code
+        // generation, even if overflow checks are disabled for the current crate. This is
+        // especially useful because overflow checks are usually disabled for the standard library.
+        AddAssign::add_assign(&mut self.count, 1);
+        Some((i, a))
+    }
+
+    #[inline]
+    default unsafe fn __iterator_get_unchecked(&mut self, idx: usize) -> Self::Item
+    where
+        Self: TrustedRandomAccess,
+    {
+        // SAFETY: the caller must uphold the contract for
+        // `Iterator::__iterator_get_unchecked`.
+        let value = unsafe { try_get_unchecked(&mut self.iter, idx) };
+        // See comment in `next()` for the reason why `Add::add()` is used here instead of `+`.
+        (Add::add(self.count, idx), value)
+    }
+
+    #[inline]
+    default fn next_back(&mut self) -> Option<Self::Item>
+    where
+        I: ExactSizeIterator + DoubleEndedIterator,
+    {
+        let a = self.iter.next_back()?;
+        let len = self.iter.len();
+        // Can safely add, `ExactSizeIterator` promises that the number of
+        // elements fits into a `usize`.
+        Some((self.count + len, a))
+    }
+}
+
+// This is the same code as above but using `intrinsics::assume()` to hint at the compiler
+// that the returned index is smaller than the length of the underlying iterator.
+//
+// This could be bound to `TrustedLen + ExactSizeIterator` or `TrustedRandomAccess` to guarantee
+// that the number of elements fits into an `usize` and that the returned length is actually the
+// real length. `TrustedRandomAccess` was selected because specialization on `ExactSizeIterator` is
+// not possible (yet?).
```

Comment:
@matthewjasper Could you comment on the possibilities here? Specializing on `ExactSizeIterator` would be useful here, but is not accepted. I see `#[rustc_specialization_trait]` and `#[rustc_unsafe_specialization_marker]` being used in some places to work around this, but I don't know in which cases those can be used.

Reply:
Considering that `TrustedLen` is unsafe + unstable + `rustc_unsafe_specialization_marker` it is more or less implied that any type that implements it is some iterator owned by the standard library. Which means there are no lifetime shenanigans that would make `ExactSizeIterator` unsafe for min_specialization.
Using that justification a `rustc_unsafe_specialization_marker` helper trait inheriting from both should be fine. Probably.

On the other hand `TrustedRandomAccess` is extending its tendrils into ever more places, so soon it should be almost equivalent in power to `TrustedLen + ExactSizeIterator`.

Reply:
`TrustedRandomAccess` seems like it would always be more specific than `TrustedLen+ExactSizeIterator`. Implementing the former on things like `BTreeMap` would probably be not a good idea, while `TrustedLen+ExactSizeIterator` would be easy.

Reply:
Then yeah, I think that should be possible. But it might be good to get a second opinion from someone who has better understanding of the min_specialization rules.

Reply:
Considering the progress that happened in the LLVM bug report and all the issues found recently involving `TrustedRandomAccess`, maybe let's just wait this out until LLVM handles it correctly?

What do you think @m-ou-se ?